Python/specialize.c

```diff
@@ -0,0 +1,165 @@
+
+#include "Python.h"
+#include "pycore_code.h"
+#include "opcode.h"
+
+
+/* We layout the quickened data as a bi-directional array:
+ * Instructions upwards, cache entries downwards.
+ * first_instr is aligned to a SpecializedCacheEntry.
+ * The nth instruction is located at first_instr[n]
+ * The nth cache is located at ((SpecializedCacheEntry *)first_instr)[-1-n]
+ * The first (index 0) cache entry is reserved for the count, to enable finding
+ * the first instruction from the base pointer.
+ * The cache_count argument must include space for the count.
+ * We use the SpecializedCacheOrInstruction union to refer to the data
+ * to avoid type punning.
+
+ Layout of quickened data, each line 8 bytes:
+
+ <cache_count>                              <---- co->co_quickened
+ <cache N-1>
+ <cache N-2>
+ ...
+ <cache 0>
+ <instr 0> <instr 1> <instr 2> <instr 3>    <--- co->co_first_instr
+ <instr 4> <instr 5> <instr 6> <instr 7>
+ ...
+ <instr N-1>
+*/
+
+static SpecializedCacheOrInstruction *
+allocate(int cache_count, int instruction_count)
+{
+    assert(sizeof(SpecializedCacheOrInstruction) == 2*sizeof(int32_t));
+    assert(sizeof(SpecializedCacheEntry) == 2*sizeof(int32_t));
+    assert(cache_count > 0);
+    assert(instruction_count > 0);
+    int count = cache_count + (instruction_count + INSTRUCTIONS_PER_ENTRY -1)/INSTRUCTIONS_PER_ENTRY;
+    SpecializedCacheOrInstruction *array = (SpecializedCacheOrInstruction *)
+        PyMem_Malloc(sizeof(SpecializedCacheOrInstruction) * count);
+    if (array == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    array[0].entry.zero.cache_count = cache_count;
+    return array;
+}
+
+static int
+get_cache_count(SpecializedCacheOrInstruction *quickened) {
+    return quickened[0].entry.zero.cache_count;
+}
+
+/* Map from opcode to adaptive opcode.
+  Values of zero are ignored. */
+static uint8_t adaptive_opcodes[256] = { 0 };
+
+/* The number of cache entries required for a "family" of instructions. */
+static uint8_t cache_requirements[256] = { 0 };
+
+/* Return the oparg for the cache_offset and instruction index,
+ * or -1 if no cache is needed or possible.
+ * The cache_offset is updated as a side-effect */
+static int
+oparg_from_instruction_and_update_offset(int index, int opcode, int original_oparg, int *cache_offset) {
+    uint8_t need = cache_requirements[opcode];
+    if (need == 0) {
+        return original_oparg;
+    }
+    assert(adaptive_opcodes[opcode] != 0);
+    int oparg = oparg_from_offset_and_index(*cache_offset, index);
+    assert(*cache_offset == offset_from_oparg_and_index(oparg, index));
+    if (oparg < 0) {
+        *cache_offset = index/2;
+        oparg = 0;
```

Comment:
what's happening here (why index/2)?

This part of the code is not currently exercised, need is always 0 above. Is it worth adding one trivial adaptive instruction and one trivial super instruction to this PR?

Reply:
I'd rather not add fake instructions. We'll be adding real ones soon enough :slightly_smiling_face:

Reply:
It's all super subtle. Here's my understanding. (Is it worth just pasting this into a comment?)

We're betting that *on average* no more than 1 in 2 instructions needs a cache entry (less for instructions that need multiple cache entries).

If it were strictly every other instruction needing a cache entry, we could find the offset of the cache entry by dividing the ordinal of the instruction in half. But of course it's not strictly like that -- sometimes you have a series of consecutive instructions that could all use a cache entry.

We don't want to have as many cache entries as instructions -- that would be quite wasteful (since 1 in 2 is still a rather high estimate). We also don't waste space (and memory access time) by having an array of pointers (IIRC that is how the existing inline cache works).

So we use the quickened instruction's oparg to compute the cache index. If we just equated the cache index with the oparg, we could have at most 256 cache entries, which is too small. So we use the oparg to compute a sort of "relative" cache index -- we take the *minimum* cache index for an instruction to be half that instruction's ordinal, and then we add oparg to that minimum value.

This function is used to compute the oparg for quickened instructions that need a cache entry (note that superinstructions don't need a cache entry). The caller is supposed to loop over all the instructions and call this function whenever an instruction qualifies for quickening.

If there's a stretch of consecutive instructions that need quickening, oparg will steadily increase. If there's a stretch of instructions that don't need quickening, the needed oparg will decrease again. If oparg were to go below zero, we adjust oparg to zero and leave some cache entries unused. (If this happens a lot we could reduce our estimate of cache entries per instruction.)

If the needed oparg would not fit in a byte we return -1, and then the caller doesn't quicken the instruction. (This should be rare; it would be interesting to have a way to report how often this happens, and if it happens a lot we should increase our estimate of cache entries per instruction. It could only happen if there are over 256 consecutive instructions that need quickening.)

Finally note that we never quicken instructions that have an extended oparg (i.e. are preceded by one or more EXTENDED_ARG pseudo-instructions). This is because we want the quickened instruction array to correspond 1-1 to the original instruction array, and checking for extended-oparg instructions is a pain. So quickening won't work very well when you have e.g. more than 256 local or free variables or reference more than 256 different globals. Such code is probably unreadable anyways. :-)

Reply:
Thanks.   This should be somewhere (comment or some doc). Maybe also worth implementing in python a simulation of this so we can get stats about how it behaves on real code. Could even go into unit tests so we get a heads up if something changes w.r.t. the assumptions as specialisations are added.

Reply:
> Maybe also worth implementing in python a simulation of this so we can get stats about how it behaves on real code.

Oooh, good idea. I have some [infrastructure](https://github.com/faster-cpython/tools/blob/main/scripts/count_opcodes.py) for this kind of thing and it would be very simple to run this against the top-N most popular PyPI packages -- once we have an idea of which opcodes could be quickened. (Maybe we could start with the opcodes that Facebook/Instagram's Shadowcode quickens, or alternatively we could look at what Pyston does.

> Could even go into unit tests so we get a heads up if something changes w.r.t. the assumptions as specialisations are added.

To qualify as a unit test it should run fast (or only be run when some flag is passed). That probably disqualifies running anything over all of PyPI (or even all of the stdlib). But a sample of the stdlib might work.

Reply:
I guess I meant a regression test - something you can run when you add specialisations to get an idea of the impact.

I’ll try to get something together once the weather permits (it’s too nice now).

Reply:
I wouldn't spend much time on that sort of analysis.
If our estimates on how much cache we need are high, we waste a very small amount of memory.
If they are low, then we don't optimize very long functions very well.
Neither are terribly important. 

As for what will need cache entries, and how large:
LOAD_GLOBAL: 2
LOAD_ATTR: 2 (Maybe 3 to handle properties and other more exotic lookups)
CALL_FUNCTION(_KW): 2
CALL_METHOD(_KW): 2
BINARY/INPLACE: 0 (Can be specialized but don't need a cache)

Reply:
I understand. From the point of view of explaining it to other people (who don't have your deep knowledge and intuition on this) - having some data is probably a more convincing than "we do this heuristic and we think it will be fine and if not it's not a big deal". 

I don't know how much that matters here.

Reply:
I did a quick measurement based on Mark's list of needed cache above, and across all the stdlib the number of cache entries per opcode is around 0.22. Over the 100 most popular PyPI packages (of a few months ago), the number is 0.3. So 0.5 is a high estimate and I expect we'll be doing fine, not wasting too much, not exceeding the limits too much.

I didn't look for long consecutive runs of cache-needing opcodes, that would be a larger project.